### PR TITLE
Vectorize PINNs and fast GPU

### PR DESCRIPTION
**Pull request by KirillZubov**

### CODE REVIEW

File: src/pinns_pde_solve.jl [hunk 1]
```diff
@@ -532,17 +532,17 @@ Base.Broadcast.broadcasted(::typeof(get_u()), cord, θ, phi) = get_u()(cord, θ,
...
         ε = εs[order]
         ε = adapt(typeof(θ),ε)
         x = adapt(typeof(θ),x)
         if order > 1
             return (derivative(phi,u,x .+ ε,εs,order-1,θ)
-                  - derivative(phi,u,x .- ε,εs,order-1,θ))*_epsilon
+                  .- derivative(phi,u,x .- ε,εs,order-1,θ)) /_epsilon
         else
-            return (u(x .+ ε,θ,phi) - u(x .- ε,θ,phi))*_epsilon
+            return (u(x .+ ε,θ,phi) .- u(x .- ε,θ,phi)) /_epsilon
```
Comment (ChrisRackauckas) [inline src/pinns_pde_solve.jl:545]: ./?

Comment (ChrisRackauckas) [reply src/pinns_pde_solve.jl:545]: you might as well fuse it.

Comment (KirillZubov) [reply src/pinns_pde_solve.jl]: done

Comment (KirillZubov) [reply src/pinns_pde_solve.jl]: @ChrisRackauckas `fastchain` doesn't work with gpu?

Comment (ChrisRackauckas) [reply src/pinns_pde_solve.jl]: it does. There's specific tests for it.

**Comment by KirillZubov** [comment]
@ChrisRackauckas vectorization led to the interruption of training with BFGS optimization

```julia
@parameters x
@variables u(..)
Dxxx = Differential(x)^3
Dx = Differential(x)

eq = Dxxx(u(x)) ~ cos(pi*x)

bcs = [u(0.) ~ 0.0,
       u(1.) ~ cos(pi),
       Dx(u(1.)) ~ 1.0]
domains = [x ∈ IntervalDomain(0.0,1.0)]

# Discretization
dx = 0.1

chain = FastChain(FastDense(1,8,Flux.σ),FastDense(8,1))
stratagy = NeuralPDE.GridTraining(dx)
discretization = NeuralPDE.PhysicsInformedNN(chain,stratagy)
pde_system = PDESystem(eq,bcs,domains,[x],[u])
prob = NeuralPDE.discretize(pde_system,discretization)
sym_prob = NeuralPDE.symbolic_discretize(pde_system,discretization)

julia> res = GalacticOptim.solve(prob, BFGS(); cb = cb, maxiters=300)
Current loss is: 2.3361920777286898
Current loss is: 2.0234650293746976
Current loss is: 1.714374496704096
Current loss is: 1.7058137441033212
Current loss is: 1.7028939668208913
Current loss is: 1.6748144828298486
Current loss is: 1.5713610560120808
Current loss is: 1.5700592309655415
Current loss is: 1.5700592309655415
```

```julia
phi = discretization.phi
analytic_sol_func(x) = (π*x*(-x+(π^2)*(2*x-3)+1)-sin(π*x))/(π^3)
xs = [domain.domain.lower:dx/10:domain.domain.upper for domain in domains][1]
u_real  = [analytic_sol_func(x) for x in xs]
u_predict  = [first(phi(x,res.minimizer)) for x in xs]
x_plot = collect(xs)
plot(x_plot ,u_real)
plot!(x_plot ,u_predict)
```
image

**Comment by ChrisRackauckas** [comment]
BFGS auto-stops when it starts to rise. Try it with `allow_f_increases=true` and setting a maximum number of iterations. Or do what I have been suggesting: 300 iterations of ADAM before BFGS.

**Comment by ChrisRackauckas** [comment]
BFGS is just too prone to hit local minima with what you are doing there, so that style of optimization code you show just shouldn't be in the tests or the tutorials.

File: docs/src/pinn/3D.md [hunk 2]
```diff
@@ -81,41 +101,17 @@ function plot_()
...
-```julia
+![3pde](https://user-images.githubusercontent.com/12683885/107345726-eefb7500-6ad4-11eb-889d-ba0fac0f1674.gif)
 
-prob = remake(prob; u0 = res.minimizer)
-res = GalacticOptim.solve(prob,ADAM(0.1);cb=cb,maxiters=1000)
 
-plot_()
-```
+## Performance benchmarks
+Here are some performance benchmarks for 3pde with various number of input points and the number of neurons in the hidden layer, measuring the time for 100 iterations. Сomparing runtime with GPU and CPU.
```
Comment (ChrisRackauckas) [inline docs/src/pinn/3D.md:115]: which GPU?

**Review by ChrisRackauckas — APPROVED** [review]
It would be good to mention the hardware that goes with the benchmarks, but that's it.

File: docs/src/pinn/3D.md (hunk 2 above)
Comment (KirillZubov) [reply docs/src/pinn/3D.md:115]:
```julia
julia> CUDA.device()
CuDevice(0): Tesla P100-PCIE-16GB
```

**Pull request merged**